src/index/hnsw/faiss_hnsw.cc

```diff
@@ -0,0 +1,1398 @@
+        //   based on the data in the future. Otherwise, data needs to be
+        //   converted into float*.
+        hnsw_index->train(rows, (const float*)data);
+
+        // done
+        index = std::move(hnsw_index);
+        return Status::success;
+    }
+};
+
+template <typename DataType>
+class BaseFaissRegularIndexHNSWFlatNodeTemplate : public BaseFaissRegularIndexHNSWFlatNode {
+ public:
+    BaseFaissRegularIndexHNSWFlatNodeTemplate(const int32_t& version, const Object& object)
+        : BaseFaissRegularIndexHNSWFlatNode(version, object, datatype_v<DataType>) {
+    }
+};
+
+namespace {
+
+// a supporting function
+expected<faiss::ScalarQuantizer::QuantizerType>
+get_sq_quantizer_type(const std::string& sq_type) {
+    std::map<std::string, faiss::ScalarQuantizer::QuantizerType> sq_types = {{"SQ6", faiss::ScalarQuantizer::QT_6bit},
+                                                                             {"SQ8", faiss::ScalarQuantizer::QT_8bit},
+                                                                             {"FP16", faiss::ScalarQuantizer::QT_fp16},
+                                                                             {"BF16", faiss::ScalarQuantizer::QT_bf16}};
+
+    // todo: tolower
+    auto itr = sq_types.find(sq_type);
+    if (itr == sq_types.cend()) {
+        return expected<faiss::ScalarQuantizer::QuantizerType>::Err(
+            Status::invalid_args, fmt::format("invalid scalar quantizer type ({})", sq_type));
+    }
+
+    return itr->second;
+}
+
+expected<bool>
+is_flat_refine(const std::optional<std::string>& refine_type) {
+    // grab a type of a refine index
+    if (!refine_type.has_value()) {
+        return true;
+    };
+
+    // todo: tolower
+    if (refine_type.value() == "FP32" || refine_type.value() == "FLAT") {
+        return true;
+    };
+
+    // parse
+    auto refine_sq_type = get_sq_quantizer_type(refine_type.value());
+    if (!refine_sq_type.has_value()) {
+        LOG_KNOWHERE_ERROR_ << "Invalid refine type: " << refine_type.value();
+        return expected<bool>::Err(Status::invalid_args, fmt::format("invalid refine type ({})", refine_type.value()));
+    }
+
+    return false;
+}
+
+// pick a refine index
+expected<std::unique_ptr<faiss::Index>>
+pick_refine_index(const DataFormatEnum data_format, const std::optional<std::string>& refine_type,
+                  std::unique_ptr<faiss::IndexHNSW>&& hnsw_index) {
+    // yes
+
+    // grab a type of a refine index
+    expected<bool> is_fp32_flat = is_flat_refine(refine_type);
+    if (!is_fp32_flat.has_value()) {
+        return expected<std::unique_ptr<faiss::Index>>::Err(Status::invalid_args, "");
+    }
+
+    const bool is_fp32_flat_v = is_fp32_flat.value();
+
+    // check input data_format
+    if (data_format == DataFormatEnum::fp16) {
+        // make sure that we're using fp16 refine
+        auto refine_sq_type = get_sq_quantizer_type(refine_type.value());
```

Comment:
Can we put this check to config? Milvus will check parameter validation through a CGO call to Knowhere in proxy for:
1. Early termination for bad request.
2. Build is an async request, it will be great to prevent bad request at beginning.

Reply:
We can put it in `CheckAndAdjust`

Reply:
Makes sense